### DISCUSSION TITLE
Documenting ToolMaker

### DISCUSSION THREAD

**Discussion by tajmone** [thoni56/ToolMaker#8] [Ideas]
@thoni56, since ToolMaker is part of the [ALAN] toolchain (i.e. it's needed to tweak the ALAN syntax), I think it's worth trying to document ToolMaker, and I'm happy to help in this respect.

You mention "the missing documentation" in the main README. Was ToolMaker never documented because it was for internal use only? or was it documented in Swedish, or the documentation simply got lost? If there was a documentation printed on paper I could scan it and digitalize it, for example. If it's in Swedish, I could create a draft translation using [DeepL], etc.

In absence of an original documentation, I could try to document the project from scratch, starting from the individual tools (which I'm currently unable to build though, see #5), and as I learn my way through start to document the notation of the grammar files, etc.

From what I gather, ToolMaker is quite an interesting set of tools, since it contains both a lexer and parser generator (whereas most of the other tools only handle the lexer or the parser). Also, while sifting through the sources I noticed it's a LALR(1) parser, so it should be similar to the [Lemon parser generator] in terms of parsing features, which is an excellent tool.

And if I'm not wrong, ToolMaker also provides some extra tools, which add value to it.

Anyhow, documenting it might be a long-term project, but it can be done nonetheless, a bit at the time, as I get to know ToolMaker in my free time (even if it's a morning I dedicate to it every so many months).

What do you think would be the best approach:

1. I start documenting it in my fork's Wiki, and whenever a document is usable I submit it to the repository.
2. I work on the documentation in a dedicated branch, and we simply merge the docs once they are production ready.

With either approach we could use Markdown or AsciiDoc, at our own discretion. Obviously, I'm more prone to use AsciiDoc since I prefer it and it would spare us porting work in case we end up creating a proper document from the various tools' READMEs.

Once I'll have access to the compiled binaries the task should be easier (I really should re-install my Ubuntu machine and just cross compile it). But since there is no documentation whatsoever right now, any input from you would be useful to set me in the right course in this task.

-------------------------------------------------------------------------------

> P.S. — Could you enable _Discussions_ and convert this Issue into a discussion? Just to keep Issues clutter-free.

[ALAN]: https://github.com/alan-if/alan "Visit the ALAN source repository"

[DeepL]: https://www.deepl.com/translator "DeepL translator website"

[Lemon parser generator]: https://github.com/tajmone/lemon-grove "Visit the Lemon Grove repository"

**Comment by thoni56** [comment]
There was extensive documentation in English. However, that has simply got lost. When I preserved this from getting lost from my "former company" (actaully a much more involved story than that) the copyright holder was still in existance. I didn't think of preserving the documentation.

I have asked around amongst my former collegues, but none of them seems to have a copy, not even the one that worked on the company taking over.

So I think the best route forward would probably be option 1). I'm guessing that the documentation will never be ready for "release", but will anyway be valuable along the way.

Agree that AsciiDoc would be the prefered format.

**Comment by tajmone** [comment]
# Updates on Documentation Wiki

@thoni56, I've managed to recover all the man page files from the repository and convert them to AsciiDoc. The only missing man page seems to be the one for the PMK tool.

Here you can find the [Toolmake(1) AsciiDoc] Wiki page, and here the [Toolmake(1) HTML Live Preview].

As you can see, I added a warning admonition about SoftLab no longer being active, and that any references to the [softlab.se] are not associated with the SoftLab Rising company which is using the web domain today (owned by a Kalle Björk; any relation to the original SoftLab?).

Before publishing the rest of the converted man pages, I wanted to make sure with you that the admonitions notes are correct.

Of course, I also need to polish the AsciiDoc sources, since the automatic conversion from man-page to AsciiDoc is not perfect and needs adjustments. But that's not going to take more than a couple of hours in total.

But I think it's already a good start, since the man pages offer some useful insights into the various tools, especially the IMP(1) man. By studying these documents, along with the repository examples and test files, I should learn my way through ToolMaker.

Obviously, a surviving copy of the _Toolmake Reference Manual_ mentioned in the man page (and possibly the other paper manuals that went along with the software product) would have been better, but let's work with what we have.





[Toolmake(1) AsciiDoc]: https://github.com/tajmone/ToolMaker/wiki/man-Toolmake
[Toolmake(1) HTML Live Preview]: https://htmlpreview.github.io/?https://raw.githubusercontent.com/wiki/tajmone/ToolMaker/man-Toolmake.html

[softlab.se]: https://softlab.se/

**Reply by thoni56** [reply]
Thanks, that's a good first step. And the admonitions about SoftLab is correct. The Swedish company SoftLab AB also had nothing to do with [SoftLab Gmbh](https://en.wikipedia.org/wiki/Softlab),

**Reply by tajmone** [reply]
> Thanks, that's a good first step. And the admonitions about SoftLab is correct. The Swedish company SoftLab AB also had nothing to do with SoftLab Gmbh,

Great! Then I've just published all the remaining man page in AsciiDoc (and HTML) on the Wiki! (I was keeping them in a local branch, waiting for confirmation):


- [IMP](https://github.com/tajmone/ToolMaker/wiki/man-IMP)
- [LMK](https://github.com/tajmone/ToolMaker/wiki/man-LMK)
- [PMK](https://github.com/tajmone/ToolMaker/wiki/man-PMK)
- [SMK](https://github.com/tajmone/ToolMaker/wiki/man-SMK)

Most man pages took only about 15 mins to port, except the IMP man page, which was both long and tricky to format due to a complex definition list with blocks, etc. But I've solved it, it just took a bit longer than expected.

The next step is to interlink the various man pages, since they cross reference each other, and possibly add some in-document xref links in the IMP man, since it's long and has various "see XXX" references.

I've annotated a couple of tweaks to be done, like ensuring that clicking on the old email shows a "dead account" pop-up, and clicking redirects to either nowhere or to your repository — I just want to avoid people contacting the company who now owns the domain, asking info about ToolMaker. You never know, they might end up receiving email with info requests, and they wouldn't be happy about it. Probably replacing the email auto-generated `mailto:` link with a link to your repo is the best solution, since it describes ToolMaker history, etc.

Once these tweaks are done, I could create a version of the man pages for the repository, e.g. in a `docs_src/` folder, and have them convert to HTML into `docs/`, so you could also serve the docs on the WWW via GHPages (in case you want to), or just have them there as part of the repo (and maybe provide Live HTML preview links). Or we could track just the ADoc sources, and expect end users (are there any?) to build the docs themselves.

**Reply by thoni56** [reply]
Nice!

One simple way to handle old references (email addresses, persons, etc.) could be to cross them over and replace with link to the repo if that is possible:

    Email to: <a href=..repo...><del>tools@softlab.se</del></a>

I think we could just track the Adoc for now, there are no users except me ;-)

It was a quite a nostalgic trip reading those pages again after almost 30 years. Thanks for bringing that back!

You mentioned that `toolmake` was not available on Windows so I got curious. I haven't had any use for it during all those years, but I noticed that `toolmake` was actually included in the repo, together with an executable for 32-bit SPARC running SYSV! Well, we have no use for that, unless you happen to have a SPARCstation laying around somewhere, so I decided to see if I can restore `toolmake, ri`pping away useless cruft like the license handling and make it run. At least to the level that you can get the `-help`  output. I'll ping you when it can be compiled.

**Reply by tajmone** [reply]
> One simple way to handle old references (email addresses, persons, etc.) could be to cross them over and replace with link to the repo if that is possible:

Yes, Asciidoctor supports the [line-through] role (aka _strikethrough_), so it's doable, and makes it clear that the email is no longer active, without redacting away the original text.

> You mentioned that `toolmake` was not available on Windows so I got curious. [...] so I decided to see if I can restore `toolmake`, ripping away useless cruft like the license handling and make it run. At least to the level that you can get the `-help` output. I'll ping you when it can be compiled.

Thanks, that's nice! It sounds like `toolmake` is useful, especially in the learning stage, since it guide the user through the process of choosing the right settings, and caters for all the basic scaffolding.

[line-through]: https://docs.asciidoctor.org/asciidoc/latest/text/text-span-built-in-roles/#built-in-roles-for-text

**Reply by thoni56** [reply]
Whoa, that's some old code! I've cleaned it up just enough to make it compile on Linux (still a lot of warnings and other uglinesses, such as K &R non-prototyped function headers...)

Here's the output

```
thoni@thoni64:~/Utveckling/ToolMaker/src/toolmake$ ./toolmake -h
toolmake -- initiator for ToolMaker tools 2.1(12) (1997-07-28 17:57)

Usage: toolmake [-[-]go|-[-]verbose|-help]... <name>

Arguments:
  <name>            -- for the subsystem

Options:
  -[-]go            -- quick toolmake with all defaults accepted (default: OFF)
  -[-]verbose       -- verbose mode (default: OFF)
  -help             -- this help
```

**Reply by thoni56** [reply]
Fun fact: the code is so old that when ToolMake generates generation dates in the generated files it becomes 19122-02-05 ;-)

I'll keep that for a while...

**Reply by tajmone** [reply]
> Fun fact: the code is so old that when ToolMake generates generation dates in the generated files it becomes 19122-02-05 

Is that related to the "Millenniums Bug"? when in 2000 it was supposed to zero-out all computer dates?

**Reply by thoni56** [reply]
Related in the sense that the writer of the code did not even think of 19 becoming 20.
